### ISSUE TITLE
Issue about the sign() and modification of the weight

### ISSUE THREAD

**Issue by blueardour** [jiecaoyu/XNOR-Net-PyTorch#72]
Hi, @jiecaoyu   Thanks for the great project and the discussion on the gradients of scaled sign function.  I've two issues when digging into the code.

1. The computing of binary weight. In the project, the binary of the weight is computing by sign() function. However, the function return 0 if the input equal 0. Would it be better for the following code:

` 
def get_binary(weight):

    binary = weight.new_ones(weight)

    binary.masked_fill_(weight <0, -1)

    return binary
   
`

2. The weight is updated by the BinOp.binarization() and BinOp.restore(). I noticed that the restored weight is modified by binarization(). Thus, if run the validate() in the main.py multiple times within the same session, the result would be different. Because every time, the restore weight is different with the pre-load weight (changed by the BinOp.restore()). 

Is my understanding correct? Thanks

**Comment by blueardour** [comment]
An additional question is: why the gradient is scaled with 1e+9 ? 
in `self.target_modules[index].grad.data = self.target_modules[index].grad.data.mul(1e+9)` ?  Sorry for the trouble.

**Comment by jiecaoyu** [comment]
Hi @blueardour 
1. There is a rare case that weights equal to 0. This is because we are using L2-Norm for regularization.

2. Yes, but the change is negligible. Because through the training process, we are calling binarization() every iteration. The meancenterConvParams() and clampConvParams() functions won't change too many weights.

3. It is used to reduce the effect of weight decay.

**Comment by blueardour** [comment]
Great, thanks for your explaination.
